flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/EliminateCrossJoinRule.java

```diff
@@ -0,0 +1,477 @@
+				.push(join.getInput(0))
+				.push(join.getInput(1))
+				.join(JoinRelType.FULL, join.getJoinFilter());
+			if (join.getPostJoinFilter() != null) {
+				relBuilder.filter(join.getPostJoinFilter());
+			}
+		} else {
+			int outerJoinCount = 0;
+			for (int i = 0; i < join.getInputs().size(); i++) {
+				if (join.getJoinTypes().get(i) != JoinRelType.INNER) {
+					outerJoinCount++;
+				}
+			}
+			Preconditions.checkState(
+				outerJoinCount <= 1,
+				"EliminateCrossJoinRule assumes that there is at most 1 outer join " +
+					"in a layer of multi-join, but " + outerJoinCount + " outer joins were found.");
+			if (outerJoinCount == 1) {
+				int numInputs = join.getInputs().size();
+				Preconditions.checkState(
+					join.getJoinTypes().get(0) == JoinRelType.RIGHT ||
+						join.getJoinTypes().get(numInputs - 1) == JoinRelType.LEFT,
+					"EliminateCrossJoinRule assumes that " +
+						"the only left outer join input must locate at the end, or" +
+						"the only right outer join input must locate at the beginning");
+			}
+
+			LoptMultiJoin loptMultiJoin = new LoptMultiJoin(join);
+
+			// try to eliminate cross join
+			Vertex joinVertexTree = multiJoinToJoinVertexTree(loptMultiJoin);
+			Mappings.TargetMapping mapping = joinVertexTreeToJoinRelTree(joinVertexTree, loptMultiJoin, relBuilder);
+
+			// apply post-join filters
+			if (join.getPostJoinFilter() != null) {
+				RexBuilder rexBuilder = join.getCluster().getRexBuilder();
+				relBuilder.filter(mapFilter(join.getPostJoinFilter(), mapping, rexBuilder));
+			}
+
+			// use projections to keep the output of the join unchanged
+			List<RexNode> projects = generateProjection(join, mapping);
+			relBuilder.project(projects);
+		}
+
+		RelNode rel = relBuilder.build();
+		call.transformTo(rel);
+	}
+
+	private Vertex multiJoinToJoinVertexTree(LoptMultiJoin multiJoin) {
+		JoinVertexTreeBuilder builder = new JoinVertexTreeBuilder(
+			multiJoin,
+			(left, right) -> {
+				boolean leftIsEqui = isEquiFilter(left.filter);
+				boolean rightIsEqui = isEquiFilter(right.filter);
+				if (leftIsEqui ^ rightIsEqui) {
+					// one of the filter is not an equi-filter
+					// equi-filter has higher priority
+					return leftIsEqui ? -1 : 1;
+				} else {
+					// both or none of the filter is an equi-filter
+					// the one with the smallest input wins
+					int a = -1;
+					int b = -1;
+					do {
+						a = left.inputBitSet.nextSetBit(a + 1);
+						b = right.inputBitSet.nextSetBit(b + 1);
+					} while (a == b && a >= 0);
+
+					if (a >= 0 && b >= 0) {
+						return a - b;
+					} else if (a < 0 && b < 0) {
+						return 0;
+					} else {
+						return a;
+					}
+				}
+			});
+
+		JoinFilter bestFilter;
+		while ((bestFilter = builder.getBestFilter()) != null) {
+			builder.innerJoin(bestFilter.inputBitSet);
+		}
+
+		return builder.toJoinVertexTree();
+	}
+
+	private boolean isEquiFilter(RexNode filter) {
+		return filter.isA(SqlKind.EQUALS) || filter.isA(SqlKind.IS_NOT_DISTINCT_FROM);
```

Comment:
does here distinguish `a=b` and `a=1` ?

Reply:
No, but after this checking we will also check if the left input and right input are both input refs. This will distinguish `a = b` from `a = 1`.